doc/flow_counters/routes_flow_counters.md

```diff
@@ -0,0 +1,475 @@
+ |:---:|:-----------:|:------------------:|-----------------------------------|
+ | 0.1 |             |      Junchao Chen  | Initial version                   |
+
+### Scope
+
+This document is the design document for route flow counter feature on SONiC.
+
+### Definitions/Abbreviations
+
+N/A
+
+### Overview
+
+Flow counters are usually used for debugging, troubleshooting and performance enhancement processes. Flow counters could cover cases like:
+
+- Routes matching the configured prefix pattern (number of hits and number of bytes)
+- Host interface traps (number of received traps per Trap ID)
+- FDB entries matching the configured VXLAN tunnel or using the VLAN ID as pattern
+- Next-Hop/Next-Hop Group/Next-Hop Group Member
+
+This document focus on route counter.
+
+### Requirements
+
+- Generic Counters shall be used as Flow Counters introduced by the feature
+- CLI shall be used for configuration, showing and clearing of statistics
+- Flow Counters for routes shall be configured using prefix patterns. The flow counter shall be bound to all routes matching the configured pattern (prefix:vrf)
+- In Phase 1 the number of configured route patterns shall be limited to 1 (enforcement shall be done during configuration via CLI)
+- In Phase 1 the number of matching routes shall be limited to the pre-configured value (default value - 30), after reboot it is not ensured that the same set of matching routes will be used for counting
+- Flow Counters shall be bound the matching routes regardless how these routes are added - manually (static) or via FRR
+- Statistics shall be configured (enabled/disabled)  and cleared using the CLI commands
+- Statistics shall be provided as a number of hit/use of a specific resource and number of bytes in packets sent via configured routes
+- When a prefix pattern is removed, Flow Counters on all routes matching the configured pattern shall be removed
+- Adding route entry shall be automatically bound to counter if counter is enabled and pattern matches
+- Removing route entry shall be automatically unbound if the entry is previously bound
+- Removing route entry shall not trigger route pattern search
+- Decreasing max allowed match count shall not trigger removing of existing bound counters
+- Increasing max allowed match count shall not trigger route pattern search
+
+### Architecture Design
+
+Flow counter shall utilize the following existing infrastructure:
+
+- Generic Counters API - defined in https://github.com/opencomputeproject/SAI/blob/master/doc/SAI-Proposal-Generic-Counters.md and already supported by the SAI layer. This feature shall support binding/unbinding of these Generic counters to/from relevant SONIC objects
+- Flex Counters framework - used for background polling and pushing the statistic information to COUNTER DB for later use (e.g. by CLI). A introduction to flex counter can be found at: https://github.com/Azure/SONiC/pull/858.
```

Comment:
Please remove vendor-specific SAI in the diagram.

Reply:
What do you mean by vendor-specific SAI?

Reply:
I see MLNX-SAI in the diagram, can we just have generic SAI modules in the diagram?